ta: pkcs11 Add implementation for Wrap/Unwrap

ruchi393 opened the pull request:
This PR adds implementation for C_Wrap() and C_KeyUnwrap().

I have done some basic testing with it using pkcs11test and I am working on adding the testcases in optee_test for the same. Since, this is a new API,I thought I would post it here to get feedback till I have the PR for optee_test ready.

Currently the implementation is as specified in PKCS#11 specification v2.40 Errata Plus Errata 01.

As I was implementing this, I found this paper which describes some pitfalls with PKCS#11 specification.
http://secgroup.dais.unive.it/wp-content/uploads/2010/10/Tookan-CCS10.pdf
In Section 5 of the paper, author has also suggested some additional things which can be done to avoid these vulnerabilities.
I plan to look at these, after adding the base testcases and getting this PR merged. 

optee_client PR related to this PR :- 
https://github.com/OP-TEE/optee_client/pull/265

optee_test PR 
https://github.com/OP-TEE/optee_test/pull/500

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/en/latest/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

ruchi393 commented:
Test case have been added and PR pushed for review - https://github.com/OP-TEE/optee_test/pull/500

Fixes done for issues found during testing.

I have also removed RFC tag from PR.  The implementation is now complete and tested as mentioned in PKCS#11 specification. Separate PR will be created for addressing the pitfalls.

@etienne-lms , @vesajaaskelainen , @ricardosalveti please help review the PR.

vesajaaskelainen commented on ta/pkcs11/include/pkcs11_ta.h:
```diff
@@ -657,6 +657,40 @@ enum pkcs11_ta_cmd {
+	 * [in]  memref[1] = wrapped key
+	 * [out] memref[2] = 32bit object handle
+	 *
+	 * This command relates to the PKCS#11 API function C_UnWrapKey().
```
`C_UnwrapKey`

vesajaaskelainen commented on ta/pkcs11/include/pkcs11_ta.h:
```diff
@@ -714,12 +748,22 @@ enum pkcs11_rc {
 	PKCS11_CKR_TOKEN_WRITE_PROTECTED	= 0x00e2,
+	PKCS11_CKR_UNWRAPPING_KEY_HANDLE_INVALID = 0x00f0,
+	PKCS11_CKR_UNWRAPPING_KEY_SIZE_RANGE	= 0x00f1,
+	PKCS11_CKR_UNWRAPPING_KEY_TYPE_INCONSISTENT = 0x00f2,
```
Missing `pkcs11_helpers.c` counter parts.

vesajaaskelainen replied:
@ruchi393 this was forgotten on latest updates.

vesajaaskelainen commented on ta/pkcs11/src/attributes.c:
```diff
@@ -276,6 +276,45 @@ bool attributes_match_reference(struct obj_attrs *candidate,
+bool attributes_match_add_reference(struct obj_attrs **head,
+				    struct obj_attrs *ref)
+{
+	size_t count = ref->attrs_count;
```
I suppose better to initialize as zero.

vesajaaskelainen commented on ta/pkcs11/src/processing_symm.c:
```diff
@@ -833,3 +833,77 @@ enum pkcs11_rc derive_key_by_symm_enc(struct pkcs11_session *session,
+		align = data_sz % TEE_AES_BLOCK_SIZE;
+		if (align) {
+			in_sz = data_sz + (TEE_AES_BLOCK_SIZE - align);
+			in_buf = TEE_Malloc(in_sz, 0);
```
Please replace `0` with `TEE_MALLOC_FILL_ZERO` so that intention is clear.

vesajaaskelainen commented on ta/pkcs11/src/processing_symm.c:
```diff
@@ -833,3 +833,77 @@ enum pkcs11_rc derive_key_by_symm_enc(struct pkcs11_session *session,
+
+	switch (proc->mecha_type) {
+	case PKCS11_CKM_AES_ECB:
+	case PKCS11_CKM_AES_CBC:
```
Sorry for the mind flow below -- kept it in here in case it helps other reviewers too.

We do have [2.14 AES Key Wrap](http://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/errata01/os/pkcs11-curr-v2.40-errata01-os-complete.html#_Toc441850517) which defines different values for mechanism as in here.

Wondering what kind of keys we should export with AES-ECB -- it is my understanding that ECB mode should not really be used for multiple blocks.

AES-CBC is slightly better.

If I look at [2.1.21 RSA AES KEY WRAP](http://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/errata01/os/pkcs11-curr-v2.40-errata01-os-complete.html#_Toc441850425) this looks somewhat better approach.

Then I was interested in how the key material is then encoded before AES.

In above:
`The recommended format for an asymmetric target key being wrapped is as a PKCS8 PrivateKeyInfo`.
And this uses keyword `recommended` which is kinda a bit nice wording in standard -- but if I am not mistaken this would say DER encode the private key material.

Then I went back to look what these mechanism here then mean.

For AES-ECB (and for CBC) it says:
`For wrapping, the mechanism encrypts the value of the CKA_VALUE attribute of the key that is wrapped, padded on the trailing end with up to block size minus one null bytes so that the resulting length is a multiple of the block size. The output data is the same length as the padded input data. It does not wrap the key type, key length, or any other information about the key; the application must convey these separately.`

And this serialization is kinda clear.

But then what makes it slightly unclear -- we have source key material (eg. key to wrap) and then we have wrapping key material (eg. RSA, AES or so key that does encryption) and then wrapping mechanism (eg. how to wrap things).

So what defines what ;)

Wrapping mechanism is easy -- that comes as argument `pMechanism` to `C_WrapKey()`.

For `CKM_AES_ECB` and `CKM_AES_CBC` that is defined clearly I would say but that can only be used for secret key objects's `CKA_VALUE` field. I suppose it should disallow any other key objects.

So I am wondering that `get_key_data_to_wrap()` probably should be subject to mechanism and used wrapping key?

Then in example if we use `CKM_RSA_AES_KEY_WRAP` and want to wrap AES key then how should that then be encoded so that it would in example be compatible with something?

For me key wrapping use case would be to transfer key from source security world to target security world. And that would require either shared secret from target security world or public key from target security world. With RSA OAEP (or RSA PKCS1.5) I kinda understand then that using public key would be the best -- but what would be the real use case with these other wrapping methods. And then the better question -- what wrapping methods should be implemented?

etienne-lms replied:
I am also a bit confused by how wrapping separates the encrypted secret value and its attributes. When referring to wrapped key blobs with HW support (on imx), didn't the key attributes were wrapped inside encrypted materials?

ruchi393 replied:
Idea of this PR was more to add boiler plate code for Wrap/Unwrap using AES-ECB and CBC mechanism. See [TS-98](https://projects.linaro.org/browse/TS-98). I did look at the CKM_AES_KEY_WRAP/WRAP_PAD but that is dependent on RFC5649 and corresponding crypto implementation is not available in TEE core. So, I left it as it is.

For the question on what defines what - below is my understanding. We get wrapping key, key to be wrapped and mechanism in C_WrapKey()

AES_ECB/CBC/AES_KEY_WRAP - things are straightforward

For CKM_RSA_AES_KEY_WRAP ->
get_key_data_to_wrap() -> key data to be wrapped 
In my understanding this depends on the target key type which is being wrapped.

This key is wrapped by an AES key -> this is a temporary key. Based on the parameters passed with CKM_RSA_AES_KEY_WRAP,  required AES key of the provided size would be created. This key is used to wrap the key data above.

Finally the RSA private key is used to wrap the AES key. Again here only value of the AES key should get wrapped.

When unwrapping, the same mechanism and parameters are passed, so unwrapping happens in reverse direction.

In the complete flow, the data which is wrapped is not dependent on the mechanism. There have been publications  that this is weak and attack prone. It is worth wrapping attributes of the key along with key data. There has been proposal for the PKCS#11 but nothing in specs as yet.

I agree with the use case you have described. Here the boiler plate code is where you have a shared secret between 2 parties. We should eventually add a mechanism like CKM_RSA_PKCS/PKCS_OEAP to this list. These seem like the mechanisms mostly used. I tried searching for usecases of mechanism - CKM_RSA_AES_KEY_WRAP but couldn't find much.

@etienne-lms , my understanding is wrapping doesn't take into account the attributes of the key as per PKCS#11 specification. This is an issue but nothing in specification on it as yet.

vesajaaskelainen commented on ta/pkcs11/src/processing.c:
```diff
@@ -747,3 +784,185 @@ enum pkcs11_rc entry_release_active_processing(struct pkcs11_client *client,
+
+	switch (get_class(key->attributes)) {
+	case PKCS11_CKO_SECRET_KEY:
+	case PKCS11_CKO_PRIVATE_KEY:
```
private keys are not supported with current code.

ruchi393 replied:
will remove

etienne-lms commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -1721,3 +1724,19 @@ enum pkcs11_rc set_key_data(struct obj_attrs **head, void *data,
+{
+	switch (get_class(obj->attributes)) {
+	case PKCS11_CKO_SECRET_KEY:
+		if (get_attribute_ptr(obj->attributes, PKCS11_CKA_VALUE, data,
```
get_attribute() expects a `void **` as 3rd arg.

I think get_key_data_to_wrap() should get a `void **data` argument and we remove the cast when this function is valled.

ruchi393 replied:
yes, you are right. I will fix this.

etienne-lms commented on ta/pkcs11/src/processing.c:
```diff
@@ -747,3 +747,165 @@ enum pkcs11_rc entry_release_active_processing(struct pkcs11_client *client,
+		goto out;
+
+	/* Set the processing state to active */
+	rc = set_processing_state(session, function, wrapping_key, NULL);
```
wrapping unwrapping are one shot functions (1single stage processing) hence no need to set and release the active processing. Call to `get_ready_session(session)` at line 803 ensures there is no pending active processing and that's enough.

ruchi393 replied:
Ok

ruchi393 replied:
Even though , we don't need to set or release active processing here, but since we later use init_symmteric_operation() which uses tee_operation etc maintained in session processing state. Hence, call to this function would be required. I realized that after I started getting crashes on removing the call to this function.

ruchi393 commented:
Comments till date addressed and PR updated

ruchi393 commented:
Ping

vesajaaskelainen commented on ta/pkcs11/include/pkcs11_ta.h:
```diff
@@ -657,6 +657,40 @@ enum pkcs11_ta_cmd {
+	 *
+	 * This command relates to the PKCS#11 API function C_WrapKey().
+	 */
+	PKCS11_CMD_WRAP_KEY = 50,
```
+1

vesajaaskelainen commented on ta/pkcs11/include/pkcs11_ta.h:
```diff
@@ -657,6 +657,40 @@ enum pkcs11_ta_cmd {
+	 *
+	 * This command relates to the PKCS#11 API function C_UnwrapKey().
+	 */
+	PKCS11_CMD_UNWRAP_KEY = 51,
```
+1

ruchi393 commented:
Rebased, squashed commits, tested again and updated the PR.

@etienne-lms , @vesajaaskelainen , please help review.

vesajaaskelainen commented on ta/pkcs11/src/pkcs11_helpers.c:
```diff
@@ -278,6 +278,9 @@ static const struct any_id __maybe_unused string_rc[] = {
 	PKCS11_ID(PKCS11_CKR_SIGNATURE_LEN_RANGE),
+	PKCS11_ID(PKCS11_CKR_UNWRAPPING_KEY_HANDLE_INVALID),
+	PKCS11_ID(PKCS11_CKR_UNWRAPPING_KEY_SIZE_RANGE),
+	PKCS11_ID(PKCS11_CKR_UNWRAPPING_KEY_TYPE_INCONSISTENT),
```
@ruchi393: please add following too here:
```
	PKCS11_CKR_WRAPPED_KEY_INVALID		= 0x0110,
	PKCS11_CKR_WRAPPED_KEY_LEN_RANGE	= 0x0112,
	PKCS11_CKR_WRAPPING_KEY_HANDLE_INVALID  = 0x0113,
	PKCS11_CKR_WRAPPING_KEY_SIZE_RANGE	= 0x0114,
	PKCS11_CKR_WRAPPING_KEY_TYPE_INCONSISTENT = 0x0115,
	PKCS11_CKR_RANDOM_SEED_NOT_SUPPORTED	= 0x0120,
	PKCS11_CKR_RANDOM_NO_RNG		= 0x0121,
```

ruchi393 replied:
Sure, I will fix it. Sorry for missing it the second time :-(

vesajaaskelainen commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -1120,7 +1120,7 @@ enum pkcs11_rc check_created_attrs_against_token(struct pkcs11_session *session,
 	 * TODO: START_DATE and END_DATE: complies with current time?
 	 */
-	return PKCS11_CKR_OK;
+	return check_access_attrs_against_token(session, head);
```
This breaks the usage pattern that has been used before.

And now there is double call to `check_access_attrs_against_token()` with: `entry_create_object()`.

Also keypair generation has the same pattern.

I would drop this commit and add it in unwrapping code as used in: `entry_create_object()`.

ruchi393 replied:
Thanks, i had missed that. You are right, I will add the additional check in unwrap/derive rather than changing check_access_attrs_against_token()

vesajaaskelainen commented on ta/pkcs11/src/processing.c:
```diff
@@ -625,8 +625,13 @@ enum pkcs11_rc entry_processing_key(struct pkcs11_client *client,
 		 * specified with C_Derive/Unwrap() in the specification. So
 		 * return the next most appropriate error.
 		 */
-		if (rc == PKCS11_CKR_KEY_FUNCTION_NOT_PERMITTED)
```
There are already similar fixups in actual implementation commit so why not have these already there?

ruchi393 replied:
Sure, I will squash these with wrap/unwrap commits.

vesajaaskelainen commented:
`Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>`
- https://github.com/OP-TEE/optee_os/pull/4469/commits/7e889650a44ec7ad35290ad18447e4b42f0fc889 - "ta: pkcs11: Allocate command ID's for wrapping/unwrapping keys"

vesajaaskelainen commented:
@ruchi393 please do check check failures.

ruchi393 commented:
Fixed issues based on the review comments.

ruchi393 commented:
@etienne-lms , @vesajaaskelainen , I have removed the build errors/warnings. Please help review.

@jbech-linaro , I am unable to figure out why IBART is failing on this. Can you please help.

jforissier commented:
@ruchi393

> I am unable to figure out why IBART is failing on this.

Unrelated to this PR, kernel build error probably caused either by the rebase of the `optee` branch onto `v5.12`, or (much more likely) this: https://github.com/OP-TEE/build/pull/474/files#diff-713ec74eedfac2d7c7c92096f8f58c262b376ee201d0ca9d448e77c348faa3a0R24

jbech-linaro commented:
Yes, after installing the packages suggested by Jerome, things seems to be back in business (last build, this PR, is OK after restarting).

vesajaaskelainen commented:
Due to other work in pkcs11 this needs a rebase.

vesajaaskelainen commented:
For commit:
- https://github.com/OP-TEE/optee_os/pull/4469/commits/0858f758694753b57cb8b3b7d025c138e83c49f3 - ta: pkcs11: Add missing error codes

`Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>`

etienne-lms commented on ta/pkcs11/src/processing.c:
```diff
@@ -747,3 +747,176 @@ enum pkcs11_rc entry_release_active_processing(struct pkcs11_client *client,
+		rc = PKCS11_CKR_KEY_HANDLE_INVALID;
+		goto out;
+	}
+
```
Commit "ta: pkcs11: Add error check when wrapping keys" should be squashed
in commit "ta: pkcs11: Add implementation for wrapping keys" here.

ruchi393 replied:
Ok

etienne-lms commented on ta/pkcs11/src/processing.c:
```diff
@@ -747,3 +747,176 @@ enum pkcs11_rc entry_release_active_processing(struct pkcs11_client *client,
+		goto out;
+	}
+
+	/* Return error if processing already active */
```
could remove this in-line comment. same for the 2 next also, IMO.

ruchi393 replied:
Ok

etienne-lms commented on ta/pkcs11/src/processing.c:
```diff
@@ -747,3 +747,176 @@ enum pkcs11_rc entry_release_active_processing(struct pkcs11_client *client,
+
+	if (get_bool(key->attributes, PKCS11_CKA_WRAP_WITH_TRUSTED) &&
+	    !get_bool(wrapping_key->attributes, PKCS11_CKA_TRUSTED)) {
+		rc = PKCS11_CKR_KEY_NOT_WRAPPABLE;
```
`DMSG("Wrap with trusted not satisfied");` ?

ruchi393 replied:
Ok

ruchi393 replied:
ok

etienne-lms commented on ta/pkcs11/src/processing_symm.c:
```diff
@@ -833,3 +833,42 @@ enum pkcs11_rc derive_key_by_symm_enc(struct pkcs11_session *session,
+		align = data_sz % TEE_AES_BLOCK_SIZE;
+		if (align) {
+			in_sz = data_sz + (TEE_AES_BLOCK_SIZE - align);
+			in_buf = TEE_Malloc(in_sz, TEE_MALLOC_FILL_ZERO);
```
If data buffer is big, we may prefer allocate only the last block.
We could address this here or in a later change.

ruchi393 replied:
I can change this to - 
                align = data_sz % TEE_AES_BLOCK_SIZE;
                if (align) {
                        if (data_sz > TEE_AES_BLOCK_SIZE) {
                                in_sz = data_sz - align;
                                res = TEE_CipherUpdate(proc->tee_op_handle,
                                                       data, in_sz, *out_buf,
                                                       out_sz);
                                if (res)
                                        return tee2pkcs_error(res);
                        } 
                        in_buf = TEE_Malloc(TEE_AES_BLOCK_SIZE,
                                                        TEE_MALLOC_FILL_ZERO);
                        if (!in_buf)
                                return PKCS11_CKR_ARGUMENTS_BAD;

                        TEE_MemMove(in_buf, (uint8_t *)data + in_sz, align);
                        in_sz = TEE_AES_BLOCK_SIZE;
                } else {
                        in_buf = data;
                        in_sz = data_sz;
                }

                res = TEE_CipherDoFinal(proc->tee_op_handle, in_buf, in_sz,
                                        *out_buf, out_sz);
                if (align)
                        TEE_Free(in_buf);

ruchi393 replied:
> > Tested out CKM_AES_CBC and seems to work.
> > Just for fun -- will try circle between OP-TEE and SoftHSM.
> 
> SoftHSMv2 seems to be a bit conservative in here.
> 
> Unwrap for devel branch (eg. not yet in release):
> 
> * wrap: https://github.com/opendnssec/SoftHSMv2/blob/develop/src/lib/SoftHSM.cpp#L6458
>   
>   * CKM_AES_KEY_WRAP -- if detected in crypto backend
>   * CKM_AES_KEY_WRAP_PAD -- if detected in crypto backend
>   * CKM_RSA_PKCS
>   * CKM_RSA_PKCS_OAEP
>   * CKM_AES_CBC
>   * CKM_AES_CBC_PAD
> * unwrap: https://github.com/opendnssec/SoftHSMv2/blob/develop/src/lib/SoftHSM.cpp#L6725
>   
>   * CKM_AES_KEY_WRAP -- if detected in crypto backend
>   * CKM_AES_KEY_WRAP_PAD -- if detected in crypto backend
>   * CKM_AES_CBC_PAD
>   * CKM_DES3_CBC_PAD
> 
> Kinda understand that DES3 is not in wrapping part.
> 
> RSA_PKCS and RSA_PKCS_OAEP I understand for wrapping but don't understand why it is not in unwrap.
> 
> For one commercial HSM in order to export AES key at all it needs to be explicitly enabled. And after enabled it barks in the log that now it was accessed.
> 
> Now I am wondering should we have mk/config.mk options to enable these weaker exports?

Sorry, I didn't exactly understand what you mean here ? Can you explain more.

vesajaaskelainen replied:
You somehow replied it in here -- I try to clarify it in "base comment"

etienne-lms commented on ta/pkcs11/src/processing.c:
```diff
@@ -538,13 +534,38 @@ enum pkcs11_rc entry_processing_key(struct pkcs11_client *client,
+	    TEE_PARAM_TYPE_GET(ptypes, 0) != TEE_PARAM_TYPE_MEMREF_INOUT ||
+	    TEE_PARAM_TYPE_GET(ptypes, 2) != TEE_PARAM_TYPE_MEMREF_OUTPUT ||
 	    out->memref.size != sizeof(obj_handle))
 		return PKCS11_CKR_ARGUMENTS_BAD;
```
I would prefer params '1 and '3 also checked. Maybe
```patch
 	if (!client ||
 	    TEE_PARAM_TYPE_GET(ptypes, 0) != TEE_PARAM_TYPE_MEMREF_INOUT ||
 	    TEE_PARAM_TYPE_GET(ptypes, 2) != TEE_PARAM_TYPE_MEMREF_OUTPUT ||
 	    out->memref.size != sizeof(obj_handle) ||
+	    TEE_PARAM_TYPE_GET(ptypes, 3) != TEE_PARAM_TYPE_NONE)
 		return PKCS11_CKR_ARGUMENTS_BAD;
 
 	switch (function) {
 	case PKCS11_FUNCTION_UNWRAP:
 		(...)
		break;
 	case PKCS11_FUNCTION_DERIVE:
+		if (TEE_PARAM_TYPE_GET(ptypes, 1) != TEE_PARAM_TYPE_NONE)
+			return PKCS11_CKR_ARGUMENTS_BAD;
+
 		operation = PKCS11_FUNCTION_ENCRYPT;
 		break;
 	default:
 		return PKCS11_CKR_ARGUMENTS_BAD;
```

ruchi393 replied:
Ok

etienne-lms commented on ta/pkcs11/src/processing.c:
```diff
@@ -647,14 +673,25 @@ enum pkcs11_rc entry_processing_key(struct pkcs11_client *client,
+						&out_buf, &out_size);
+			break;
+		default:
+			rc = PKCS11_CKR_ARGUMENTS_BAD;
```
prefer `rc = PKCS11_CKR_GENERAL_ERROR;`
Maybe also `assert(0)` or even `TEE_Panic(PKCS11_CKR_GENERAL_ERROR)`.

ruchi393 replied:
Changed and also added a TEE_Panic(function);

ruchi393 replied:
Changed

etienne-lms reviewed (commented):
Comments for commits "ta: pkcs11: Add implementation for wrapping keys"
and "ta: pkcs11: Add implementation for unwrapping keys".

ruchi393 commented:
Rebased on master
Tag applied for 0858f75 - ta: pkcs11: Add missing error codes
Review comments addressed

ruchi393 commented:
@etienne-lms , @vesajaaskelainen , sorry for the late update. I have addressed the review comments. Please help review.

vesajaaskelainen commented:
I will add this to my devel branch before my commits and play with it.

vesajaaskelainen commented on ta/pkcs11/src/processing.h:
```diff
@@ -124,4 +127,12 @@ enum pkcs11_rc generate_ec_keys(struct pkcs11_attribute_head *proc_params,
+enum pkcs11_rc unwrap_key_by_symm(struct pkcs11_session *session, void *data,
+				  uint32_t data_sz, void **out_buf,
+				  uint32_t *out_sz);
+
```
These should be in "symmetric" section in the file

ruchi393 replied:
Sure, I will move it there. Didn't notice it while rebasing.

vesajaaskelainen commented on ta/pkcs11/src/pkcs11_attributes.h:
```diff
@@ -198,4 +198,7 @@ enum pkcs11_rc set_key_data(struct obj_attrs **head, void *data,
 enum pkcs11_rc add_missing_attribute_id(struct obj_attrs **pub_head,
 					struct obj_attrs **priv_head);
 
+enum pkcs11_rc get_key_data_to_wrap(struct pkcs11_object *obj, void **data,
```
There is set_key_data method... so why not  move this next to it?

ruchi393 replied:
Ok

etienne-lms commented on ta/pkcs11/src/processing.c:
```diff
@@ -1003,3 +1003,185 @@ enum pkcs11_rc entry_release_active_processing(struct pkcs11_client *client,
+
+		session->processing->mecha_type = proc_params->id;
+
+		rc  = wrap_data_by_symm_enc(session, key_data, key_sz, &out_buf,
```
why passing `&out_ptr`, not `out_ptr` value straight?

ruchi393 replied:
Will change it.

etienne-lms reviewed (commented):
Comments for "ta: pkcs11: Add implementation for wrapping keys".

etienne-lms commented on ta/pkcs11/src/processing.c:
```diff
@@ -903,14 +933,26 @@ enum pkcs11_rc entry_processing_key(struct pkcs11_client *client,
+						&out_buf, &out_size);
+			break;
+		default:
+			rc = PKCS11_CKR_GENERAL_ERROR;
```
minor: can remove this line since `TEE_Panic()` below.

ruchi393 replied:
Ok

etienne-lms reviewed (commented):
Commit "ta: pkcs11: Add implementation for unwrapping keys" looks good to me aside a minor comment.

etienne-lms reviewed (commented):
Comments for "ta: pkcs11: Add support for indirect templates".

ruchi393 commented on ta/pkcs11/src/processing.c:
```diff
@@ -1003,3 +1003,185 @@ enum pkcs11_rc entry_release_active_processing(struct pkcs11_client *client,
+	 * disallow this.
+	 */
+	if (key == wrapping_key) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
```
Spec doesn't talk about this error case explicitly, so I used PKCS11_CKR_ARGUMENTS_BAD over here. I can change it to CKR_WRAPPING_KEY_HANDLE_INVALID if you think that is more appropriate. I don't have any specific preference here.

ruchi393 commented on ta/pkcs11/src/processing.c:
```diff
@@ -1003,3 +1003,185 @@ enum pkcs11_rc entry_release_active_processing(struct pkcs11_client *client,
+		goto out;
+	}
+
+	rc = set_processing_state(session, function, wrapping_key, NULL);
```
Ok, I will remove the set/release

ruchi393 replied:
Unfortunately we can' remove this. Currently set_processing_state does much more than setting the state as it initializing proc. I think a cleanup may be required to separate this out.

etienne-lms replied:
addressed

vesajaaskelainen commented:
Oh... If you do changes please to fixup them on commits where they should be. I prefer to review them as "final" versions.

I continue tomorrow on this.

etienne-lms commented:
> Oh... If you do changes please to fixup them on commits where they should be. I prefer to review them as "final" versions.

Ok, I'm fine with that. @ruchi393, if you can squash straight fixup commits, it's fine.

ruchi393 commented on ta/pkcs11/src/processing.c:
```diff
@@ -1003,3 +1003,185 @@ enum pkcs11_rc entry_release_active_processing(struct pkcs11_client *client,
+		break;
+	case PKCS11_CKO_PRIVATE_KEY:
+		/* Key type not supported as yet */
+		rc = PKCS11_CKR_GENERAL_ERROR;
```
Ok.

ruchi393 commented on ta/pkcs11/src/processing.c:
```diff
@@ -1003,3 +1003,185 @@ enum pkcs11_rc entry_release_active_processing(struct pkcs11_client *client,
+
+	rc = get_key_data_to_wrap(key, &key_data, &key_sz);
+	if (rc) {
+		rc = PKCS11_CKR_KEY_NOT_WRAPPABLE;
```
Ok

ruchi393 commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -871,6 +890,13 @@ create_attributes_from_template(struct obj_attrs **out, void *template,
+	 * is not consistent with the object class or type and return error.
+	 * In current implementation such attributes are ignored and not
+	 * added to final object.
+	 */
```
IIUC, sanitize_client_object will just remove duplicates from temp , sanitize it and create a temporary template "temp". Further when attrs are created using create_*(), the attributes in temp which are not valid for a particular object type will be removed. Ideally if this temp contains an attribute which is not consistent with the class or type, error should be returned. 

So, I would say this comment needs to remain here.

etienne-lms replied:
If  attributes in `temp` are not consistent with the `temp` class/type, `create_*_attributes()` will simply not consider these attributes and the created object will not contain them.
Do I miss somthing?

ruchi393 commented on ta/pkcs11/src/attributes.c:
```diff
@@ -276,6 +276,45 @@ bool attributes_match_reference(struct obj_attrs *candidate,
+					   pkcs11_ref.size);
+			if (rc)
+				return false;
+			candidate = *head;
```
Changed

ruchi393 commented on ta/pkcs11/src/attributes.c:
```diff
@@ -276,6 +276,45 @@ bool attributes_match_reference(struct obj_attrs *candidate,
+			rc = add_attribute(head, pkcs11_ref.id,
+					   ref_attr + sizeof(pkcs11_ref),
+					   pkcs11_ref.size);
+			if (rc)
```
Yes, you are right. I will change the return type to CKR_ value

ruchi393 commented on ta/pkcs11/src/processing_symm.c:
```diff
@@ -833,3 +833,42 @@ enum pkcs11_rc derive_key_by_symm_enc(struct pkcs11_session *session,
+			in_sz = data_sz + (TEE_AES_BLOCK_SIZE - align);
+			in_buf = TEE_Malloc(in_sz, TEE_MALLOC_FILL_ZERO);
+			if (!in_buf)
+				return PKCS11_CKR_ARGUMENTS_BAD;
```
Ok

ruchi393 commented on ta/pkcs11/src/processing.c:
```diff
@@ -1003,3 +1003,185 @@ enum pkcs11_rc entry_release_active_processing(struct pkcs11_client *client,
+	uint32_t key_sz = 0;
+	void *out_buf = params[2].memref.buffer;
+	uint32_t out_size = params[2].memref.size;
+	enum processing_func function = PKCS11_FUNCTION_WRAP;
```
ok

ruchi393 commented:
> > Oh... If you do changes please to fixup them on commits where they should be. I prefer to review them as "final" versions.
> 
> Ok, I'm fine with that. @ruchi393, if you can squash straight fixup commits, it's fine.

Sure, I can do that. I have made fixes based on comments till date, squashed them with the commits and updated PR.

vesajaaskelainen commented:
In Commit "ta: pkcs11: Add missing error codes" -- Signed-off-by and reviewed-by are in reverse order. Convention seems to be that Sob is the last one.

jforissier commented:
> In Commit "ta: pkcs11: Add missing error codes" -- Signed-off-by and reviewed-by are in reverse order. Convention seems to be that Sob is the last one.

I have to say I have never really paid attention to that, and in fact I mostly hace the opposite habit... (author S-o-b followed by any R-b/A-b, that is everything in more or less chronological order). Any pointers/examples?

ruchi393 commented:
I am seeing some crashes currently with the changes in PR. @vesajaaskelainen , please wait before picking this up on your devel branch. I am debuugging those. Sorry for this. I thought changes were trivial but it seems I did something wrong which is causing a crash,

vesajaaskelainen commented:
> I am seeing some crashes currently with the changes in PR. @vesajaaskelainen , please wait before picking this up on your devel branch. I am debuugging those. Sorry for this. I thought changes were trivial but it seems I did something wrong which is causing a crash,

Thanks for the comment as I just merged them to my tree and compiled them ;)

I'll do something else in meanwhile.

vesajaaskelainen commented:
> > In Commit "ta: pkcs11: Add missing error codes" -- Signed-off-by and reviewed-by are in reverse order. Convention seems to be that Sob is the last one.
> 
> I have to say I have never really paid attention to that, and in fact I mostly have the opposite habit... (author S-o-b followed by any R-b/A-b, that is everything in more or less chronological order). Any pointers/examples?

Hmm.. Now that you said opposite I started looking at differences.

I seems to be in part of doing it in order leaving author sob as last but there seems to be very varying practices. So parhaps we can ignore my comment on the matter.

ruchi393 commented:
Issue fixed. @vesajaaskelainen , you can pick up the branch now.

vesajaaskelainen commented:
I have been lately playing with [python-pkcs11](https://python-pkcs11.readthedocs.io/) library on my yocto based test environment.

Testing key warpping I started with simple:
```python
    wrapped_key_ciphertext = wrapping_key.wrap_key(secret_key)
```

And this now results with:
```
59: C_WrapKey
2021-06-30 06:18:45.396
[in] hSession = 0x1
pMechanism->type=(null)
[in] hWrappingKey = 0x2
[in] hKey = 0x1
Returned:  7 CKR_ARGUMENTS_BAD
```

Now is there some universal default to use when mechanism is defined as NULL? 

With SoftHSMv2 2.6.1 this works fine without an issue.

vesajaaskelainen commented:
And after setting up mechanism type to `AES_ECB`...

This seems to need the same `CKR_ARGUMENTS_BAD` fix too to `libckteec`. Eg. the query for the needed buffer size without first setting variable value to zero for return value (eg. leave it to whatever is in stack).

I have fix in queue for encryption/decryption and just made test fix for wrapping.

Tested with the `libckteec` fix and reran the test app and now it passes and input and output keys are match.

@ruchi393 do you want that I provide the fix for `libckteec` and then you would provide the test for the fix in test suite to test that `NULL, NULL` is checked for CKR_ARGUMENTS_BAD and then similar tests for size query `NULL, &len` for `CKR_BUFFER_TOO_SMALL`?

ruchi393 commented:
> And after setting up mechanism type to `AES_ECB`...
> 
> This seems to need the same `CKR_ARGUMENTS_BAD` fix too to `libckteec`. Eg. the query for the needed buffer size without first setting variable value to zero for return value (eg. leave it to whatever is in stack).
> 
> I have fix in queue for encryption/decryption and just made test fix for wrapping.
> 
> Tested with the `libckteec` fix and reran the test app and now it passes and input and output keys are match.
> 
> @ruchi393 do you want that I provide the fix for `libckteec` and then you would provide the test for the fix in test suite to test that `NULL, NULL` is checked for CKR_ARGUMENTS_BAD and then similar tests for size query `NULL, &len` for `CKR_BUFFER_TOO_SMALL`?
Thanks @vesajaaskelainen . Please provide the fix in libckteec. I will add corresponding tests.

ruchi393 commented on ta/pkcs11/src/processing_symm.c:
```diff
@@ -833,3 +833,52 @@ enum pkcs11_rc derive_key_by_symm_enc(struct pkcs11_session *session,
+				in_sz = data_sz - align;
+				res = TEE_CipherUpdate(proc->tee_op_handle,
+						       data, in_sz, out_buf,
+						       out_sz);
```
My bad. I totally forgot  to deal with the out_buf. Will make the correction. Thanks for catching this.

vesajaaskelainen commented:
Tested out CKM_AES_CBC and seems to work.

Just for fun -- will try circle between OP-TEE and SoftHSM.

vesajaaskelainen commented:
> Tested out CKM_AES_CBC and seems to work.
> 
> Just for fun -- will try circle between OP-TEE and SoftHSM.

SoftHSMv2 seems to be a bit conservative in here.

Unwrap for devel branch (eg. not yet in release):
- wrap: https://github.com/opendnssec/SoftHSMv2/blob/develop/src/lib/SoftHSM.cpp#L6458
  - CKM_AES_KEY_WRAP -- if explicitly enabled
  - CKM_AES_KEY_WRAP_PAD -- if explicitly enabled
  - CKM_RSA_PKCS
  - CKM_RSA_PKCS_OAEP
  - CKM_AES_CBC
  - CKM_AES_CBC_PAD
- unwrap: https://github.com/opendnssec/SoftHSMv2/blob/develop/src/lib/SoftHSM.cpp#L6725
  - CKM_AES_KEY_WRAP -- if explicitly enabled
  - CKM_AES_KEY_WRAP_PAD -- if explicitly enabled
  - CKM_AES_CBC_PAD
  - CKM_DES3_CBC_PAD

Kinda understand that DES3 is not in wrapping part.

Why AES_KEY_WRAPS need to be explicitly enabled while keeping AES_CBC/AES_CBC_PAD in there always available don't understand.

RSA_PKCS and RSA_PKCS_OAEP I understand for wrapping but don't understand why it is not in unwrap.

For one commercial HSM in order to export AES key at all it needs to be explicitly enabled. And after enabled it barks in the log that now it was accessed.

Now I am wondering should we have mk/config.mk options to enable these weaker exports?

ruchi393 commented on ta/pkcs11/src/attributes.h:
```diff
@@ -216,11 +216,15 @@ static inline enum pkcs11_rc get_u32_attribute(struct obj_attrs *head,
 				struct obj_attrs *candidate);
 
+/*
+ * Return PKCS11_CKR_OK on success, or a PKCS11 return code.
```
Ok

ruchi393 commented on ta/pkcs11/src/processing.c:
```diff
@@ -903,14 +933,25 @@ enum pkcs11_rc entry_processing_key(struct pkcs11_client *client,
 		 * key. Hence pass function as PKCS11_FUNCTION_ENCRYPT
 		 * to init_symm_operation()
```
Ok

ruchi393 commented:
Changes done as per @etienne-lms comments , patches squashed and force pushed.

vesajaaskelainen commented:
> > > Tested out CKM_AES_CBC and seems to work.
> > > Just for fun -- will try circle between OP-TEE and SoftHSM.
> > 
> > SoftHSMv2 seems to be a bit conservative in here.
> > 
> > Unwrap for devel branch (eg. not yet in release):
> > 
> > * wrap: https://github.com/opendnssec/SoftHSMv2/blob/develop/src/lib/SoftHSM.cpp#L6458
> >   
> >   * CKM_AES_KEY_WRAP -- if detected in crypto backend
> >   * CKM_AES_KEY_WRAP_PAD -- if detected in crypto backend
> >   * CKM_RSA_PKCS
> >   * CKM_RSA_PKCS_OAEP
> >   * CKM_AES_CBC
> >   * CKM_AES_CBC_PAD
> > * unwrap: https://github.com/opendnssec/SoftHSMv2/blob/develop/src/lib/SoftHSM.cpp#L6725
> >   
> >   * CKM_AES_KEY_WRAP -- if detected in crypto backend
> >   * CKM_AES_KEY_WRAP_PAD -- if detected in crypto backend
> >   * CKM_AES_CBC_PAD
> >   * CKM_DES3_CBC_PAD
> > 
> > Kinda understand that DES3 is not in wrapping part.
> > 
> > RSA_PKCS and RSA_PKCS_OAEP I understand for wrapping but don't understand why it is not in unwrap.
> > 
> > For one commercial HSM in order to export AES key at all it needs to be explicitly enabled. And after enabled it barks in the log that now it was accessed.
> > 
> > Now I am wondering should we have mk/config.mk options to enable these weaker exports?

@ruchi393> Sorry, I didn't exactly understand what you mean here ? Can you explain more.

Well I could not test out the cross functionality because we did not have matching unwrap operation.

ECB seems to have not been implemented at all on SoftHSMv2 for key wrapping and I am wondering is it because of weakness in the ECB mode itself. In the mode when you encrypt more than one block it is not safe anymore. Also it is missing IV.

Modes with padding is probably a good idea to implement in later PRs.

AES_KEY_WRAP* is probably better when comparing against ECB.

So second wondering in there was about what kind of default modes we should support -- as we have set some stricter rules for some operations should this also be subject to those so that eg. one can disable ECB key wrapping support so if one has PKCS#11 operation access then they could not try to export existing keys with ECB out and try to hack em.

Eg. we have `CFG_PKCS11_TA_ALLOW_DIGEST_KEY` should we have similar setting like `CFG_PKCS11_TA_ALLOW_KEY_WRAP_AES_ECB`? And then the better question is whether it should be enabled by default or not.

vesajaaskelainen commented:
> I have been lately playing with [python-pkcs11](https://python-pkcs11.readthedocs.io/) library on my yocto based test environment.
> 
> Testing key warpping I started with simple:
> 
> ```python
>     wrapped_key_ciphertext = wrapping_key.wrap_key(secret_key)
> ```
> 
> And this now results with:
> 
> ```
> 59: C_WrapKey
> 2021-06-30 06:18:45.396
> [in] hSession = 0x1
> pMechanism->type=(null)
> [in] hWrappingKey = 0x2
> [in] hKey = 0x1
> Returned:  7 CKR_ARGUMENTS_BAD
> ```
> 
> Now is there some universal default to use when mechanism is defined as NULL?
> 
> With SoftHSMv2 2.6.1 this works fine without an issue.

Back to this one after the fixes.

The default is defined in `python-pkcs11` and not in the specification (at least did not find anything on that):
https://github.com/danni/python-pkcs11/blob/c6c8efe099c1a05f680d67fb9d6f3ea90191eeb8/pkcs11/defaults.py#L77

In the device with OP-TEE I see now:
```
63: C_WrapKey
2021-07-02 09:44:51.412
[in] hSession = 0x1
pMechanism->type=(null)
[in] hWrappingKey = 0x2
[in] hKey = 0x1
Returned:  112 CKR_MECHANISM_INVALID
```
Which does not help much -- but if I do the same trace with Ubuntu+SoftHSMv2 the story is a bit nicer:
```
59: C_WrapKey
2021-07-02 12:48:48.016
[in] hSession = 0x1
pMechanism->type=CKM_AES_KEY_WRAP             
[in] hWrappingKey = 0x3
[in] hKey = 0x2
[out] pWrappedKey[*pulWrappedKeyLen] NULL [size : 0x28 (40)]
Returned:  0 CKR_OK
```

And as we do not yet support `CKM_AES_KEY_WRAP` then that is fine.

Then I updated `opensc` from 0.19.0 -> 0.21.0 and now I also see this in device:
```
63: C_WrapKey
2021-07-02 10:01:19.568
[in] hSession = 0x1
pMechanism->type=CKM_AES_KEY_WRAP             
[in] hWrappingKey = 0x2
[in] hKey = 0x1
Returned:  112 CKR_MECHANISM_INVALID
```

So the `(null)` was only a problem in older version of `opensc`.

vesajaaskelainen commented:
If I am not mistaken the `(PKCS11_)CKM_AES_CBC_PAD` for encryption (and wrapping) is broken currently. It does not add actual padding. TEE Internal Core API requires that the padding is done by the caller and thus provides only algo `TEE_ALG_AES_CBC_NOPAD` and PKCS#11 says that it will handle padding internally from caller -- thus we are supposed to be doing that.

vesajaaskelainen commented on ta/pkcs11/src/processing.c:
```diff
@@ -1003,3 +1003,181 @@ enum pkcs11_rc entry_release_active_processing(struct pkcs11_client *client,
+	struct pkcs11_object *key = NULL;
+	struct obj_attrs *req_attrs = NULL;
+	uint32_t wrapping_key_handle = 0;
+	uint32_t key_handle = 0;
```
`CK_INVALID_HANDLE` is for the purpose of telling invalid handle.

etienne-lms replied:
0 is the invalid handle identifier as per _ta/pkcs11/src/handle.c_ implementation.
I'm fine with 0 as is.
We can create a specific P-R to `#define HANDLE_INVALID 0` in _handle.h_ and use it in  ta/pkcs11 source files.

ruchi393 replied:
Leaving as it is for now.

etienne-lms commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -1865,3 +1868,19 @@ enum pkcs11_rc add_missing_attribute_id(struct obj_attrs **pub_head,
+			return PKCS11_CKR_ARGUMENTS_BAD;
+		break;
+	default:
+		return PKCS11_CKR_GENERAL_ERROR;
```
This should not happen, right? What about an `assert(0)`?

ruchi393 replied:
Ok

etienne-lms commented on ta/pkcs11/src/processing.c:
```diff
@@ -1003,3 +1003,181 @@ enum pkcs11_rc entry_release_active_processing(struct pkcs11_client *client,
+		goto out;
+
+	rc = get_attribute_ptr(wrapping_key->attributes,
+			       PKCS11_CKA_WRAP_TEMPLATE, (void *)&req_attrs,
```
is the cast needed?

ruchi393 replied:
Compiler complains if cast is not put.

_ta/pkcs11/src/processing.c: In function ‘entry_wrap_key’:
ta/pkcs11/src/processing.c:1149:37: error: passing argument 3 of ‘get_attribute_ptr’ from incompatible pointer type [-Werror=incompatible-pointer-types]
 1149 |           PKCS11_CKA_WRAP_TEMPLATE, &req_attrs, &size);
      |                                     ^~~~~~~~~~
      |                                     |
      |                                     struct obj_attrs **
In file included from ta/pkcs11/src/processing.c:14:
ta/pkcs11/src/attributes.h:134:13: note: expected ‘void **’ but argument is of type ‘struct obj_attrs **’
  134 |      void **attr_ptr, uint32_t *attr_size);
      |      ~~~~~~~^~~~~~~~
cc1: all warnings being treated as errors_

vesajaaskelainen replied:
I suppose the right cast would then be `(void **)`

ruchi393 replied:
Fixed.

ruchi393 replied:
I was too quick to add the fix. This results in a compiler error - dereferencing type-punned pointer might break strict-aliasing rules [-Werror=strict-aliasing]

I think the best way to do this is to change the type of req_attrs to void * which will avoid all these casts. I will update the PR.

etienne-lms commented on ta/pkcs11/src/processing_symm.c:
```diff
@@ -833,3 +833,68 @@ enum pkcs11_rc derive_key_by_symm_enc(struct pkcs11_session *session,
+				res = TEE_CipherUpdate(proc->tee_op_handle,
+						       data, in_sz, tmp_buf,
+						       &tmp_sz);
+				tmp_buf = tmp_buf + tmp_sz;
```
nitpicking: would prefer `tmp_buf += tmp_sz`

ruchi393 replied:
Ok

etienne-lms reviewed (commented):
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for the 2 first commits:
"ta: pkcs11: Allocate command ID's for wrapping/unwrapping keys"
"ta: pkcs11: Add missing error codes"

Comments for commit "ta: pkcs11: Add implementation for wrapping keys".

etienne-lms commented on ta/pkcs11/src/processing.c:
```diff
@@ -794,12 +790,53 @@ enum pkcs11_rc entry_processing_key(struct pkcs11_client *client,
+	uint32_t in_size = 0;
 	void *out_buf = NULL;
 	uint32_t out_size = 0;
+	enum processing_func operation = PKCS11_FUNCTION_ENCRYPT;
```
prefer `PKCS11_FUNCTION_UNKNOWN`.

ruchi393 replied:
Ok

etienne-lms reviewed (commented):
Few comments for commit "ta: pkcs11: Add implementation for unwrapping keys".

vesajaaskelainen commented on ta/pkcs11/src/processing_symm.c:
```diff
@@ -833,3 +833,68 @@ enum pkcs11_rc derive_key_by_symm_enc(struct pkcs11_session *session,
+				tmp_sz = *out_sz - tmp_sz;
+			}
+			in_buf = TEE_Malloc(TEE_AES_BLOCK_SIZE,
+							TEE_MALLOC_FILL_ZERO);
```
whitespace issues here

ruchi393 replied:
Ok

vesajaaskelainen commented on ta/pkcs11/src/processing_symm.c:
```diff
@@ -833,3 +833,68 @@ enum pkcs11_rc derive_key_by_symm_enc(struct pkcs11_session *session,
+			if (!in_buf)
+				return PKCS11_CKR_DEVICE_MEMORY;
+
+			TEE_MemMove(in_buf, (uint8_t *)data + in_sz, align);
```
unnecessary cast.

etienne-lms commented on ta/pkcs11/src/attributes.c:
```diff
@@ -276,6 +276,41 @@ bool attributes_match_reference(struct obj_attrs *candidate,
+			if (rc)
+				return rc;
+		} else if (rc || !value || size != pkcs11_ref.size ||
+		   TEE_MemCompare(ref_attr + sizeof(pkcs11_ref), value, size)) {
```
nitpicking: prefer a strict standard indentation:
```c
		} else if (rc || !value || size != pkcs11_ref.size ||
			   TEE_MemCompare(ref_attr + sizeof(pkcs11_ref), value,
					  size)) {
			return PKCS11_CKR_TEMPLATE_INCONSISTENT;
		}
```
or
```c
		} else {
			if (rc || !value || size != pkcs11_ref.size ||
			    TEE_MemCompare(ref_attr + sizeof(pkcs11_ref), value,
					   size))
				return PKCS11_CKR_TEMPLATE_INCONSISTENT;
		}
```

ruchi393 replied:
Ok

etienne-lms commented on ta/pkcs11/src/attributes.h:
```diff
@@ -215,12 +215,18 @@ static inline enum pkcs11_rc get_u32_attribute(struct obj_attrs *head,
 				struct obj_attrs *candidate);
 
+/*
+ * Check if attributes from @ref are all found in @head and if not are these
```
"... and if not **~~are~~* these are added..."
Shorter other suggestion:
` * Check attributes from @ref are all found or added into @head`

ruchi393 replied:
Changed to " Check attributes from @ref are all found or added into @head"

etienne-lms commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -757,6 +757,9 @@ create_attributes_from_template(struct obj_attrs **out, void *template,
 	uint32_t mechanism_id = PKCS11_CKM_UNDEFINED_ID;
+	struct obj_attrs *req_attrs = NULL;
+	uint32_t size = 0;
+	uint32_t indirect_template = PKCS11_CKA_DERIVE_TEMPLATE;
```
would prefer `PKCS11_CKA_UNDEFINED_ID`.

ruchi393 replied:
Ok

etienne-lms commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -831,6 +834,21 @@ create_attributes_from_template(struct obj_attrs **out, void *template,
+			rc = attributes_match_add_reference(&temp, req_attrs);
+			if (rc)
+				goto out;
+		}
```
maybe `assert(rc == PKCS11_CKR_OK || rc == PKCS11_RV_NOT_FOUND);`

ruchi393 replied:
Why assert over here ?

vesajaaskelainen commented on ta/pkcs11/src/processing_symm.c:
```diff
@@ -833,3 +833,68 @@ enum pkcs11_rc derive_key_by_symm_enc(struct pkcs11_session *session,
+			if (!in_buf)
+				return PKCS11_CKR_DEVICE_MEMORY;
+
+			TEE_MemMove(in_buf, (uint8_t *)data + in_sz, align);
```
Wrapping can have any length input data.

If input data size is < AES block size this will overflow.

In example if data_sz == 8 -> data[16 .. 23] is being accessed.

etienne-lms replied:
you're right. We should add:
```patch
 			if (data_sz > TEE_AES_BLOCK_SIZE) {
 				in_sz = data_sz - align;
 				(...)
 				tmp_sz = *out_sz - tmp_sz;
+			} else {
+				in_sz = 0;
 			}
 			in_buf = TEE_Malloc(TEE_AES_BLOCK_SIZE, TEE_MALLOC_FILL_ZERO);
 			if (!in_buf)
 				return PKCS11_CKR_DEVICE_MEMORY;
 
 			TEE_MemMove(in_buf, (uint8_t *)data + in_sz, align);
```

ruchi393 replied:
Done..

etienne-lms reviewed (commented):
Comments for commit "ta: pkcs11: Add support for indirect templates".

vesajaaskelainen commented on ta/pkcs11/src/processing_symm.c:
```diff
@@ -900,7 +900,7 @@ enum pkcs11_rc wrap_data_by_symm_enc(struct pkcs11_session *session,
 
 enum pkcs11_rc unwrap_key_by_symm(struct pkcs11_session *session, void *data,
-				  uint32_t data_sz, void *out_buf,
+				  uint32_t data_sz, void **out_buf,
```
These pointer-pointer changes should be in commit that originates them.

ruchi393 replied:
Changed

vesajaaskelainen commented on ta/pkcs11/src/processing_symm.c:
```diff
@@ -898,3 +898,38 @@ enum pkcs11_rc wrap_data_by_symm_enc(struct pkcs11_session *session,
+						data_sz, out_buf, out_sz);
+		}
+		if (tee2pkcs_error(res)) {
+			TEE_Free(*out_buf);
```
add:
```c
*out_buf = NULL;
```
So that we do not have possible double free issue.

ruchi393 replied:
Done

vesajaaskelainen commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -1865,3 +1868,19 @@ enum pkcs11_rc add_missing_attribute_id(struct obj_attrs **pub_head,
 		return set_attribute(pub_head, PKCS11_CKA_ID, id2, id2_size);
 }
+
+enum pkcs11_rc get_key_data_to_wrap(struct pkcs11_object *obj, void **data,
```
Please move this right after `set_key_data()`

ruchi393 replied:
Ok

ruchi393 commented:
> > > > Tested out CKM_AES_CBC and seems to work.
> > > > Just for fun -- will try circle between OP-TEE and SoftHSM.
> > > 
> > > 
> > > SoftHSMv2 seems to be a bit conservative in here.
> > > Unwrap for devel branch (eg. not yet in release):
> > > 
> > > * wrap: https://github.com/opendnssec/SoftHSMv2/blob/develop/src/lib/SoftHSM.cpp#L6458
> > >   
> > >   * CKM_AES_KEY_WRAP -- if detected in crypto backend
> > >   * CKM_AES_KEY_WRAP_PAD -- if detected in crypto backend
> > >   * CKM_RSA_PKCS
> > >   * CKM_RSA_PKCS_OAEP
> > >   * CKM_AES_CBC
> > >   * CKM_AES_CBC_PAD
> > > * unwrap: https://github.com/opendnssec/SoftHSMv2/blob/develop/src/lib/SoftHSM.cpp#L6725
> > >   
> > >   * CKM_AES_KEY_WRAP -- if detected in crypto backend
> > >   * CKM_AES_KEY_WRAP_PAD -- if detected in crypto backend
> > >   * CKM_AES_CBC_PAD
> > >   * CKM_DES3_CBC_PAD
> > > 
> > > Kinda understand that DES3 is not in wrapping part.
> > > RSA_PKCS and RSA_PKCS_OAEP I understand for wrapping but don't understand why it is not in unwrap.
> > > For one commercial HSM in order to export AES key at all it needs to be explicitly enabled. And after enabled it barks in the log that now it was accessed.
> > > Now I am wondering should we have mk/config.mk options to enable these weaker exports?
> 
> @ruchi393> Sorry, I didn't exactly understand what you mean here ? Can you explain more.
> 
> Well I could not test out the cross functionality because we did not have matching unwrap operation.
> 
> ECB seems to have not been implemented at all on SoftHSMv2 for key wrapping and I am wondering is it because of weakness in the ECB mode itself. In the mode when you encrypt more than one block it is not safe anymore. Also it is missing IV.
> 
> Modes with padding is probably a good idea to implement in later PRs.

I agree

> 
> AES_KEY_WRAP* is probably better when comparing against ECB.

I am in agreement with you on this. I wanted to start with this algorithm, however underlying TEE core implementation doesn't support them ( AES key Wrap - RFC 3394 and AES key Wrap PAD is RFC5649), so I started wit the easier ones.

> 
> So second wondering in there was about what kind of default modes we should support -- as we have set some stricter rules for some operations should this also be subject to those so that eg. one can disable ECB key wrapping support so if one has PKCS#11 operation access then they could not try to export existing keys with ECB out and try to hack em.
> 
> Eg. we have `CFG_PKCS11_TA_ALLOW_DIGEST_KEY` should we have similar setting like `CFG_PKCS11_TA_ALLOW_KEY_WRAP_AES_ECB`? And then the better question is whether it should be enabled by default or not.

I like the idea. We can add CFG_PKCS11_TA_ALLOW_KEY_WRAP_AES_ECB and I would say lets disable it by default. As there are already known weakness with this algorithm, if someone wants to enable it they should do it consciously. Once this PR is merged, i can create a PR for this.

ruchi393 commented on ta/pkcs11/src/processing.c:
```diff
@@ -794,12 +790,53 @@ enum pkcs11_rc entry_processing_key(struct pkcs11_client *client,
+			return PKCS11_CKR_ARGUMENTS_BAD;
+
+		/*
+		 * Some unwrap mechanism require encryption to be
```
ok

ruchi393 commented:
Comments addressed and PR updated.

vesajaaskelainen commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -1821,6 +1824,23 @@ enum pkcs11_rc set_key_data(struct obj_attrs **head, void *data,
 	}
 }
 
+enum pkcs11_rc get_key_data_to_wrap(struct pkcs11_object *obj, void **data,
```
Now looking this in place -- all neighbors has type `struct obj_attrs **head` and not `struct pkcs11_object *`. So perhaps it would be better to change this one too like that.

ruchi393 replied:
Changed

vesajaaskelainen commented on ta/pkcs11/src/processing.c:
```diff
@@ -1003,3 +1003,181 @@ enum pkcs11_rc entry_release_active_processing(struct pkcs11_client *client,
+		params[2].memref.size = out_size;
+
+out:
+	release_active_processing(session);
```
This requires own label as there are goto out's that have yet to have claimed processing session for this method.

And after processing is claimed all remaining goto's should then be targetting that new label.

ruchi393 replied:
Changed.

vesajaaskelainen commented on ta/pkcs11/src/processing.c:
```diff
@@ -1003,3 +1003,181 @@ enum pkcs11_rc entry_release_active_processing(struct pkcs11_client *client,
+
+		session->processing->mecha_type = proc_params->id;
+
+		rc  = wrap_data_by_symm_enc(session, key_data, key_sz, out_buf,
```
remove extra space `rc__=`

ruchi393 replied:
done

vesajaaskelainen commented on ta/pkcs11/src/processing_symm.c:
```diff
@@ -833,3 +833,71 @@ enum pkcs11_rc derive_key_by_symm_enc(struct pkcs11_session *session,
+enum pkcs11_rc wrap_data_by_symm_enc(struct pkcs11_session *session,
+				     void *data, uint32_t data_sz,
+				     void *out_buf, uint32_t *out_sz)
+
```
remove extra empty line

ruchi393 replied:
done

vesajaaskelainen commented on ta/pkcs11/src/processing_symm.c:
```diff
@@ -833,3 +833,71 @@ enum pkcs11_rc derive_key_by_symm_enc(struct pkcs11_session *session,
+		if (align) {
+			if (data_sz > TEE_AES_BLOCK_SIZE) {
+				in_sz = data_sz - align;
+				res = TEE_CipherUpdate(proc->tee_op_handle,
```
Missing error checking for this.

ruchi393 replied:
If we get TEE_ERROR_SHORT_BUFFER , I think we can add and assert(0), as this should not happen as we have taken care of that check in line 857. If there is any other error, we should return. I will add a check.

vesajaaskelainen commented on ta/pkcs11/src/processing_symm.c:
```diff
@@ -901,3 +901,39 @@ enum pkcs11_rc wrap_data_by_symm_enc(struct pkcs11_session *session,
+		res = TEE_CipherDoFinal(proc->tee_op_handle, data, data_sz,
+					out_buf, out_sz);
+		if (res == TEE_ERROR_SHORT_BUFFER) {
+			*out_buf = TEE_Malloc(*out_sz, 0);
```
Please use `TEE_MALLOC_FILL_ZERO` for clarity.

ruchi393 replied:
Changed

vesajaaskelainen commented:
For commit:
- https://github.com/OP-TEE/optee_os/pull/4469/commits/63691ea26c2a067557b271a411ece69a9b5a41c9 ta: pkcs11: Add access check on new object when deriving/unwrapping

`Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>`

ruchi393 commented:
Updated the PR with comments addressed

vesajaaskelainen commented:
For commits:
- https://github.com/OP-TEE/optee_os/pull/4469/commits/56cb0e73e5d71dc79ded58fe2e207643dd7b1ea8 ta: pkcs11: Add implementation for wrapping keys
- https://github.com/OP-TEE/optee_os/pull/4469/commits/ea53281a572193faaa38d80b80a30d01c783286e ta: pkcs11 Add implementation for Wrap/Unwrap
- https://github.com/OP-TEE/optee_os/pull/4469/commits/4507415ebd9ed440620c08e5da469f0f10709d5c ta: pkcs11: Add support for indirect templates

`Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>`

etienne-lms reviewed (commented):
remaining comments for commits "ta: pkcs11: Add implementation for wrapping keys"
and "ta: pkcs11: Add implementation for unwrapping keys".

`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits
"ta: pkcs11: Add support for indirect templates"
and "ta: pkcs11: Add access check on new object when deriving/unwrapping".

vesajaaskelainen commented:
Just for fun -- did quick & dirty implementation for CKM_AES_CBC_PAD in my devel branch and then took latest develop branch from softhsm which does have the matching support and wrapped secret with op-tee and then unwrapped in softhsm and resulting secret key matched.

ruchi393 commented:
Fixed based on the comments , PR updated.

ruchi393 reviewed (commented):
> For commits:
> 
> * [56cb0e7](https://github.com/OP-TEE/optee_os/commit/56cb0e73e5d71dc79ded58fe2e207643dd7b1ea8) ta: pkcs11: Add implementation for wrapping keys
> * [ea53281](https://github.com/OP-TEE/optee_os/commit/ea53281a572193faaa38d80b80a30d01c783286e) ta: pkcs11 Add implementation for Wrap/Unwrap
> * [4507415](https://github.com/OP-TEE/optee_os/commit/4507415ebd9ed440620c08e5da469f0f10709d5c) ta: pkcs11: Add support for indirect templates
> 
> `Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>`

Thanks @vesajaaskelainen . Tags applied.

ruchi393 commented:
> remaining comments for commits "ta: pkcs11: Add implementation for wrapping keys"
> and "ta: pkcs11: Add implementation for unwrapping keys".
> 
> `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits
> "ta: pkcs11: Add support for indirect templates"
> and "ta: pkcs11: Add access check on new object when deriving/unwrapping".

Thanks @etienne-lms . Tags applied and PR modified based on last comments.

jforissier commented:
@etienne-lms would you like to Ack the wrap and unwrap commits? This looks ready for merging before -rc1.

etienne-lms commented on ta/pkcs11/src/processing.c:
```diff
@@ -1003,3 +1003,181 @@ enum pkcs11_rc entry_release_active_processing(struct pkcs11_client *client,
+	 */
+	if (key == wrapping_key) {
+		rc = PKCS11_CKR_WRAPPING_KEY_HANDLE_INVALID;
+		goto out_free;
```
s/out_free/out/

etienne-lms reviewed (commented):
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits
"ta: pkcs11: Add implementation for wrapping keys"
and "ta: pkcs11: Add implementation for unwrapping keys".

ruchi393 commented:
Tags applied. Thanks @etienne-lms and @vesajaaskelainen for the review.
